### DISCUSSION TITLE
AIO should not block installation if its own domain name resolves to a private IP address during setup

### DISCUSSION THREAD

**Discussion by Lanchon** [nextcloud/all-in-one#3699] [Questions] [answered]
AIO is too picky about validating its environment during setup: it blocks installation if its own domain name resolves to a private IP address, and this is a perfectly sane configuration in some setups.

### example setup in which a private IP address makes sense

- the network in which you plan to install AIO is behind a router/internet gateway which you can configure but which you cannot replace (eg. it is the ISP's gateway and it does not support bridge mode; it always NATs).
- you setup domain `nc.my.com` to point to the WAN interface of the router.
- you setup router port-forwards for the relevant ports (80, 443, 8443).

however this router only forwards said ports if the traffic arriving at the WAN interface originates from the WAN. LAN traffic destined for the WAN interface is discarded and there is nothing you can do about it. (i would say this is, unfortunately, a rather common setup.)

to cope with this situation:

- you manage your own local DNS service. (you were doing this already to be able to access you LAN's DHCP clients by name, to use DNS over TLS/HTTPS, DNSSEC, and cache DNS responses.)
- you add a host entry to your DNS server that points `nc.my.com` to the internal IP of server running AIO.

this way:
- when outside your LAN, you resolve `nc.my.com` to the router's WAN interface and you reach AIO though the port forwards.
- when inside your LAN, you resolve `nc.my.com` to the server running AIO and reach it directly.

however this setup fails because of the AIO setup bug i am reporting: **AIO is too picky about validating its environment during setup and it blocks installation if its own domain name resolves to a private IP address.**

### workaround and proof that this is indeed an AIO setup bug

to workaround this i had to:
- temporarily disable the `nc.my.com` host entry in the local DNS service.
- add a destination NAT rule to the AIO server's nftables:
  - rewrite packets destined to the router's WAN IP and route them to localhost.
- setup AIO.
- remove the DNAT rule.
- reinstate the `nc.my.com` entry.

proof of this being a setup bug:
- after successful AIO setup and undoing of the above workaround, AIO works flawlessly form inside and outside the LAN, proving that the routable IP requirement during setup is overly picky.

### how to solve the bug

either:
- remove the public IP check
- or trigger a warning instead of blocking the install
- or at least provide a way to bypass the check (not recommended):
  - if it already exists, include how-to in the displayed error message, because i could not find it
  - it this is done using an env var, document it in the AIO setup page (i do not want to start installing AIO to find that it has this egregious and unexpected IP resolving requirement half-way during the install)

### additional info

why is the DNAT rule workaround not an acceptable setup?
- the router's WAN IP is dynamic, which means that the DNAT rule would need to be updated each time the router's lease is renewed.
- you need to add an intervening router for all exit traffic in the LAN, or else you need to add the DNAT rules in all the hosts of the LAN (and most of which, eg. phones, will not allow them).
- either the DNAT rule rewrites all traffic, which would mean that all services in the LAN need to be hosted in the same server, or you need DNAT rules for each protocol/port pair, duplicating the router's port forward setup (with the added complication that these entries must be modified with each DHCP lease renewal).
  - in comparison, for the local host entry solution you just need to define a name for each service, eg. the `nc.` part in `nc.my.com`, and add a host entry for each service (routing to different internal servers if needed).

---

### Host OS 
Linux Mint 21.2 Victoria base: Ubuntu 22.04 jammy
Linux 6.2.0-36-generic x86_64

#### Nextcloud AIO version 
Nextcloud AIO v7.5.1

#### Current channel 
latest

**Comment by szaimen** [comment, marked as answer]
Hi, see https://github.com/nextcloud/all-in-one#how-to-skip-the-domain-validation

**Reply by Lanchon** [reply]
thank you. i had read that before installing and i disregarded it... with reason.

you see, when i read the readme, "skip domain validation" did not mean anything at all to me. the reason is simple: AFAICT, there is no mention of the requirement that the nextcloud domain resolves to a public IP, and there is no mention that this will be validated during install. so there is no way the user will know what "skip domain validation" is all about. and if you read something you don't have a clue about what it is talking, you of course just wasted your time and didn't learn anything.

in my view, this issue stands as an issue and should not have been converted to a discussion.

alternatives to fix this issue:
1. disable the validation.
2. make it a warning instead of a fatal error.
3. document the public IP requirement for installation somewhere and before the install command is given. move the `SKIP_DOMAIN_VALIDATION` text to that paragraph or else point to it from there.

right now the only way to know what "skip domain validation" is about, is trying to install AIO and fail at it.

at a minimum, there is a documentation issue. but i'd prefer solution 1 or 2: i don't really see why you need to block installation of AIO in an intranet for example. do one thing and do it well. AIO cannot help you setup networks, so don't take half-baked shots at it either.

and since this is now a discussion: i really, really don't see the point of the validation and `SKIP_DOMAIN_VALIDATION` at all. why add a flag to make AIO do what you tell it to do? what is it helping?

**Reply by videvelop** [reply]
Totally agree with @Lanchon 's 1, 2 as solutions .  Domain validation is not at all required in intranet setup.
